[JIT] Propagate profiled information to DifferentiableGraph outputs

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78875

Without profiled outputs, autodiff can't tell whether or not the outputs of a DifferentiableGraph should requires_grad. Autodiff would default to requires_grad=True if there was no profiled information, causing autodiff to mark tensors as requires_grad when they shouldn't have. This adds requires_grad info onto the type of the output, if it can be found in later uses of the output.

Adds a test for correct autodiff requires_grad behavior and also a test to make sure the output type is correctly annotated in create_autodiff_subgraphs.

davidberard98 committed:
[JIT] Propagate profiled information to DifferentiableGraph outputs

Without profiled outputs, autodiff can't tell whether or not the outputs of a DifferentiableGraph should requires_grad. Autodiff would default to requires_grad=True if there was no profiled information, causing autodiff to mark tensors as requires_grad when they shouldn't have. This adds requires_grad info onto the type of the output, if it can be found in later uses of the output.

Adds a test for correct autodiff requires_grad behavior and also a test to make sure the output type is correctly annotated in create_autodiff_subgraphs.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -301,6 +301,89 @@ class SubgraphSlicer {
+    c10::optional<bool> requiresGrad = c10::nullopt;
+    for (auto& use : diff_graph->output(i)->uses()) {
+      if (use.user->kind() == prim::profile) {
+        requiresGrad = getProfileNodeRequiresGrad(use.user);
```
If a value has multiple uses, and one of the the uses isn't profiled, we don't want to overwrite the existing use's requires_grad. 
```
if (use.user->kind() == prim::profile) {
   if (auto req_grad_use = getProfileNodeRequiresGrad(use.user)) {
      requiresGrad = req_grad_use;
   }
}
```

eellison commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -301,6 +301,89 @@ class SubgraphSlicer {
+          if (dg_use.user->kind() == prim::profile) {
+            requiresGrad = getProfileNodeRequiresGrad(dg_use.user);
+            if (requiresGrad) {
+              break;
```
nit: we are doing the early break for `if (requiresGrad)` in this loop but not the loop above

eellison commented on torch/csrc/jit/passes/create_autodiff_subgraphs.cpp:
```diff
@@ -301,6 +301,89 @@ class SubgraphSlicer {
+        Value* dg_value = dg->inputs()[use.offset];
+        for (auto& dg_use : dg_value->uses()) {
+          if (dg_use.user->kind() == prim::profile) {
+            requiresGrad = getProfileNodeRequiresGrad(dg_use.user);
```
maybe also do the same don't overwrite check as above

eellison reviewed (approved):
Nice !!!

davidberard98 committed:
Update on "[JIT] Propagate profiled information to DifferentiableGraph outputs"


Without profiled outputs, autodiff can't tell whether or not the outputs of a DifferentiableGraph should requires_grad. Autodiff would default to requires_grad=True if there was no profiled information, causing autodiff to mark tensors as requires_grad when they shouldn't have. This adds requires_grad info onto the type of the output, if it can be found in later uses of the output.

Adds a test for correct autodiff requires_grad behavior and also a test to make sure the output type is correctly annotated in create_autodiff_subgraphs.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Propagate profiled information to DifferentiableGraph outputs"


Without profiled outputs, autodiff can't tell whether or not the outputs of a DifferentiableGraph should requires_grad. Autodiff would default to requires_grad=True if there was no profiled information, causing autodiff to mark tensors as requires_grad when they shouldn't have. This adds requires_grad info onto the type of the output, if it can be found in later uses of the output.

Adds a test for correct autodiff requires_grad behavior and also a test to make sure the output type is correctly annotated in create_autodiff_subgraphs.

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot help

davidberard98 commented:
@pytorchbot revert -m "Internal failures were bisected to this change" -c ghfirst
